src/deque.rs

```diff
@@ -811,6 +811,189 @@ impl<T, S: VecStorage<T> + ?Sized> DequeInner<T, S> {
+                // In either case, if a 0 desired length is provided,
+                // `back` will equal `front` and `full` will be unset,
+                // properly marking the deque as empty.
+                self.full = false;
+
+                // If `drop_front` causes a panic, the Dropper will still be called to drop it's slice during unwinding.
+                // In either case, front will always be dropped before back.
+                let _back_dropper = Dropper(&mut *drop_back);
+                ptr::drop_in_place(drop_front);
+            }
+        }
+    }
+
+    /// Retains only the elements specified by the predicate.
+    ///
+    /// In other words, remove all elements `e` for which `f(&e)` returns false.
+    /// This method operates in place, visiting each element exactly once in the
+    /// original order, and preserves the order of the retained elements.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use heapless::Deque;
+    ///
+    /// let mut buf: Deque<_, 10> = Deque::new();
+    /// buf.extend(1..5);
+    /// buf.retain(|&x| x % 2 == 0);
+    /// assert_eq!(buf.make_contiguous(), [2, 4]);
+    /// ```
+    ///
+    /// Because the elements are visited exactly once in the original order,
+    /// external state may be used to decide which elements to keep.
+    ///
+    /// ```
+    /// use heapless::Deque;
+    ///
+    /// let mut buf: Deque<_, 10> = Deque::new();
+    /// buf.extend(1..6);
+    ///
+    /// let keep = [false, true, true, false, true];
+    /// let mut iter = keep.iter();
+    /// buf.retain(|_| *iter.next().unwrap());
+    /// assert_eq!(buf.make_contiguous(), [2, 3, 5]);
+    /// ```
+    pub fn retain<F>(&mut self, mut f: F)
+    where
+        F: FnMut(&T) -> bool,
+    {
+        self.retain_mut(|elem| f(elem));
+    }
+
+    /// Retains only the elements specified by the predicate.
+    ///
+    /// In other words, remove all elements `e` for which `f(&mut e)` returns false.
+    /// This method operates in place, visiting each element exactly once in the
+    /// original order, and preserves the order of the retained elements.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use heapless::Deque;
+    ///
+    /// let mut buf: Deque<_, 10> = Deque::new();
+    /// buf.extend(1..5);
+    /// buf.retain_mut(|x| {
+    ///     if *x % 2 == 0 {
+    ///         *x += 1;
+    ///         true
+    ///     } else {
+    ///         false
+    ///     }
+    /// });
+    /// assert_eq!(buf.make_contiguous(), [3, 5]);
+    /// ```
+    pub fn retain_mut<F>(&mut self, mut f: F)
+    where
+        F: FnMut(&mut T) -> bool,
+    {
+        let len = self.storage_len();
+        let mut idx = 0;
+        let mut cur = 0;
+
+        // Stage 1: Check if all values can be retained.
+        while cur < len {
+            // Safety: `cur` must be under the deque's total length as per the loop condition
+            if !f(unsafe { self.get_unchecked_mut(cur) }) {
```

Comment:
Why deviate from `std`'s implementation and use unsafe here? I expect the bound check introduced in the safe path to be optimized out.

Reply:
Fair! My reasoning came from the fact that Deque doesn't impl `ops::IndexMut` (and most things in heapless seem to not, I'm unsure how intentional of a choice that is still). I like to avoid `unwrap`s in `unsafe` blocks for a few reasons; `expect` is _fine_, but I also felt it was unnecessarily explicit? I mean, `std::VecDeque`'s `IndexMut` impl comes out to be the same as what I have now changed it to, a `get_mut()` with an `except()`.

So maybe another PR adding those `ops` impls would be in order? Unless this was an explicit choice to not allow indexing like that, I'm still new to heapless as a whole :sweat_smile:.